datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1056,6 +1071,94 @@ private void onDatastreamChange(List<DatastreamGroup> datastreamGroups) {
+            .collect(Collectors.toList()).contains(dgName)).forEach(obsoleteDs -> _adapter.cleanUpPartitionMovement(connectorType, obsoleteDs));
+
+        // Filtered all live datastreamGrou as we process only datastream which have
+        // both partition assignment info and the target assignment
+        List<DatastreamGroup> toProcessedDatastreamGroups =
+            liveDatastreamGroups.stream().filter(group2 -> toMoveDatastream.contains(group2.getName()))
+                .filter(group3 -> datastreamPartitions.keySet().contains(group3.getName()))
+                .collect(Collectors.toList());
+
+        for (DatastreamGroup dg : toProcessedDatastreamGroups) {
+          DatastreamGroupPartitionsMetadata subscribedPartitions = connectorInstance.getDatastreamPartitions().get(dg.getName())
+              .orElseThrow(() -> new DatastreamTransientException("partition listener is not ready yet for datastream " + dg.getName()));
+          Map<String, Set<String>> suggestedAssignment = _adapter.getPartitionMovement(dg.getConnectorName(), dg.getName());
+          assignmentByInstance = strategy.movePartitions(assignmentByInstance, suggestedAssignment,
+              subscribedPartitions);
+          toCleanup.add(dg);
+        }
+      }
+
+      for (String key : assignmentByInstance.keySet()) {
+        newAssignmentsByInstance.put(key, new ArrayList<>(assignmentByInstance.get(key)));
+      }
+
+      _adapter.updateAllAssignments(newAssignmentsByInstance);
+
+      //clean up stored target assignment after the assignment is updated
+      for (DatastreamGroup dg : toCleanup) {
+        _adapter.cleanUpPartitionMovement(dg.getConnectorName(), dg.getName());
+      }
+
+      _log.info("Partition movement completed: datastreamGroup, assignment {} ", assignmentByInstance);
+
+      shouldRetry = false;
+    } catch (DatastreamTransientException ex) {
+      _log.warn("Partition movement failed, retry again after a configurable period", ex);
+      shouldRetry = true;
+    } catch (Exception ex) {
+      //We do not retry if it is not transient exception
+      _log.error("Partition movement failed, Exception: ", ex);
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, "handleLeaderPartitionMovement", NUM_ERRORS, 1);
+
+    }
+    if (!shouldRetry) {
+      _adapter.cleanupOldUnusedTasks(previousAssignmentByInstance, newAssignmentsByInstance);
+      updateCounterForMaxPartitionInTask(newAssignmentsByInstance);
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, NUM_PARTITION_MOVEMENTS, 1);
+    }  else if (!leaderPartitionMovementScheduled.get()) {
+      _log.info("Schedule retry for leader movement tasks");
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, "handleLeaderPartitionMovement", NUM_RETRIES, 1);
+      leaderPartitionMovementScheduled.set(true);
+      _executor.schedule(() -> {
+        _eventQueue.put(CoordinatorEvent.createPartitionMovementEvent());
+        leaderPartitionMovementScheduled.set(false);
```

Comment:
Didn't quite understand the purpose of this flag - shouldn't leaderPartitionMovementScheduled be set to true since partition movement event is added to the queue (i.e. movement is scheduled) just before this statement?

Reply:
I actually follow the convention that we currently did in LEADER_DO_ASSIGNMENTS. I think it is to avoid multiple one be put in the executor which schedules the event. But it is not used by queue as the queue itself will handle dedup.